design/runtime-parameters.md

```diff
@@ -0,0 +1,226 @@
+They're rendered like a plain `object` in the UI.
+
+| YAML string type | Notes |
+|------------------|-------|
+| `steps` | sequence of steps
+| `jobs` | sequence of jobs
+| `stages` | sequence of stages
+
+## Display hints
+
+Optionally, the pipeline author can hint to the UI what widget should be displayed.
+Not all widgets are compatible with all data types, and UI may evolve over time, so these are only hints.
+
+| Display hint | Compatible data types | Default for |
+|--------------|-----------------------|-------------|
+| `oneline` | string, number | string, number
+| `multiline` | string, object | object
+| `pickList` | number, boolean, enum | enum
+| `checkbox` | boolean | boolean
+| `radio` | number, boolean, enum | -
+
+For data types not listed in the above table, there is a "special" UI (such as a file picker, pool picker, etc.) that's automatically used.
+
+*This part can be postponed. It's nice-to-have, especially for cases like single line vs multiline input.*
+
+## Full examples
+
+### Dial in the amount of parallelism
+
+```yaml
+parameters:
+- name: parallelism
+  displayName: How much parallelism
+  type: number
+  default: 5
+  values:
+  - 1
+  - 5
+  - 10
+
+job: Test
+strategy:
+  parallel: ${{ parameters.parallelism }}
+steps:
+- script: echo Running one leg of tests...
+```
+ 
+### Select configs to run
+
+```yaml
+parameters:
+- name: configs
+  type: string
+  default: x86,x64
+
+jobs:
+- ${{ if contains(parameters.configs, 'x86') }}:
+  - job: x86
+    steps:
+    - script: echo Building x86...
+- ${{ if contains(parameters.configs, 'x64') }}:
+  - job: x64
+    steps:
+    - script: echo Building x64...
+- ${{ if contains(parameters.configs, 'arm') }}:
+  - job: arm
+    steps:
+    - script: echo Building arm...
+```
+
+*There's a possible future enhancement hiding here: a data type of "flags" which would be like an enum but allow multiple specification.
+This version is more verbose but also clear and straightforward.*
+
+### Select your pool
+
+```yaml
+parameters:
+- name: pool
+  type: pool
+  default: ubuntu-16.04
+
+job: Build
+pool:
+  vmImage: ${{ parameters.pool }}
+steps:
+- script: echo Building on the pool of your choice
+```
+
+### Connecting parameters with strong typing
+
+This is common for sophisticated customers such as .NET.
+
+```yaml
+# template.yml - comes from the centralized infrastructure team
+
+parameters:
+- name: customerSteps
+  type: steps
+
+steps:
+- script: echo Pre-step injected on all jobs
+- ${{ each customerStep in parameters.customerSteps }}:
+  - ${{ customerStep }}
+- script: echo Post-step injected on all jobs
+```
+
+When a client repo such as Roslyn goes to consume this template:
+
+```yaml
+# azure-pipelines.yml - in the Roslyn repo
+
+resources:
+  repositories:
+  - repository: arcade
+    name: dotnet/arcade
+
+steps:
+- template: template.yml@arcade
+  parameters:
```

Comment:
I'm not following this example. What happens here?

Reply:
The Arcade team offers a template which expects a list of steps. The Roslyn team uses that template and passes in their list of steps.

Today, it's completely untyped. Arcade has to blindly assume that Roslyn (and other customers) gave valid YAML for steps. It will fail with mysterious errors if they don't. With this change, we will know in the parser that Arcade expected a list of steps, and if the YAML isn't right, we can give a better error. It also opens the door to offering editor support: when the consumer uses a template that expects a list of steps, we'll be able to IntelliSense and syntax-highlight them correctly.

Reply:
Ah, I read this set of parameters as run parameters not template parameters.